### ISSUE TITLE
Autocorrect: Replace a typed word with a pre-defined word after pressing space. ( "im" with "I'm")

### ISSUE THREAD

**Issue by smartfonreddit** [AnySoftKeyboard/AnySoftKeyboard#1684]
Hello,

Type `im` and hit enter. It should ideally auto-correct  to `I'm,` but it doesn't. I've tried adding/linking the two words in the app's and Android's personal dictionaries, but it doesn't help. Same with `whats` and `what's`. This little detail makes typing really slow.

Suggestion is to have a dictionary where you can link im to I'm. When you type im and hit SPACE, it should automatically convert it to I'm without having to manually reach the suggestion bar on the top.

Android 9. OnePlus 6T

**Comment by appletalk** [comment]
This is a very frustrating issue.  I cannot seem to train the keyboard to do otherwise.  This occurs for many contractions in English.  Id correcting to ID rather than I'd is incorrect and slows typing considerably.

**Comment by JonnyHaystack** [comment]
Corrections like these should really be made default, but it would also be helpful to be able to create manual autocorrect mappings. I also have the issue that I type "Youre" and it autocorrects to "you're", thus incorrectly starting the sentence with a lower case.

**Comment by TheShanMan** [comment]
Also contractions seem to be unsupported by glide. Contractions so ubiquitous in English that this is essential for glide.

**Comment by gabrc52** [comment]
Other keyboards (like the AOSP keyboard and Gboard) don't have the apostrophe button, so contractions are done automatically. 

Even if there's an apostrophe button, I think that contractions should be autocorrected.

**Comment by gabrc52** [comment]
To add to this, some bottom rows in the settings don't even have the apostrophe button -- how are you supposed to type words like "don't" with them?

**Comment by gabrc52** [comment]
While manual autocorrections may be useful, I believe they're not the solution. The solution is to simply autocorrect as most would expect -- "Im" to "I'm", "Youre" to "You're", etc. If manual autocorrections are implemented, it would mean almost everyone who doewnloads AnySoftKeyboard would add them in a similar way.

**Comment by gabrc52** [comment]
After some more usage, I can see that AnySoftKeyboard **does** autocorrect some words like "dont" to "don't". However, it fails for example in "sont" where it autocorrects it to "song" instead of "don't".

I think the problem is that "im" and "whats" are actually in its dictionary so it believes they are _correct_ without taking into account that you probably didn't intend to type "im" or "whats".

I've described this behavior in more detail in issue #2821.

**Comment by TheShanMan** [comment]
Yeah unfortunately if you don't start on the right letter, you're screwed. It'll only give you words that start with the actual letter you started on. While gboard on the other hand is much more tolerant of start points. Plus ASK's correct suggestions are words that are close in the dictionary, not close in swiping pattern. I'm glad ASK exists since I decided not to use gboard anymore, but boy, do I hope these things get improved.

But back on point, I'm not sure it has ever _auto_corrected to a contraction for me. In some cases it does show them in the list of suggested alternatives though. And in other cases they aren't even in the list.

**Comment by gabrc52** [comment]
> boy, do I hope these things get improved.

Yeah, it's also one of the main annoyances I have, and I think I've found what makes autocorrections *kinda off*. Could you help in adding a thumbs up to issue #2821?

> But back on point, I'm not sure it has ever auto corrected to a contraction for me.

I also believed it didn't autocorrect contractions at all, but I just tried, and it does correct "dont" to "don't" and "youre" to "you're".
